internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -0,0 +1,117 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package customresource provides functionality to upload Copilot custom resources.
+package asset
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+)
+
+type filterAction int
+
+const (
+	include filterAction = iota + 1
+	exclude
+)
+
+// UploadFunc is the function signature to upload contents to a destination.
+type UploadFunc func(dest string, contents io.Reader) (url string, err error)
+
+// UploadInput is the input of Upload.
+type UploadInput struct {
+	Source      string
+	Destination string
+	Includes    []string
+	Excludes    []string
+	Upload      UploadFunc
```

Comment:
nit: Can we include doc comments for these fields 🥺 I assume that source is an absolute path but destination is a location in S3?